client/client.go

```diff
@@ -803,53 +780,148 @@ func (c *Client) Call(ctx context.Context, procedure string, options wamp.Dict,
+		}
 
-			var args wamp.List
-			var kwargs wamp.Dict
-			var err error
+		return msg, nil
+	case *wamp.Error:
+		return nil, RPCError{msg, procedure}
+	default:
+		return nil, unexpectedMsgError(msg, wamp.RESULT)
+	}
+}
 
-			// Now need to check ppt_serializer (in pair with ppt_scheme)
-			// and deserialize payload with appreciate serializer
-			if pptScheme == WampPPTScheme {
-				args, kwargs, err = unpackE2EEPayload(msg.Details, msg.Arguments)
-			} else {
-				args, kwargs, err = unpackPPTPayload(msg.Details, msg.Arguments)
+// CallProgressive makes progressive calls to the procedure corresponding to
+// the given URI.
+//
+// @see full description for Call method above. The only difference is that this
+// method returns a callback that client code may call to pass a progressive chunks
+// of data to the Callee. Also, please note that all progressive results (if they
+// will be at all) will be returned using progcb callback and the final result will be
+// returned using finalrescb callback
+//
+// This function returns a callback function that should be called with payload data chunks
+// To finish a progressive call the last callback MUST be called with isLast = true
+func (c *Client) CallProgressive(ctx context.Context, procedure string, options wamp.Dict, args wamp.List, kwargs wamp.Dict, finalrescb ResultHandler, progcb ProgressHandler) (SendProgressiveData, error) { //nolint:lll
+	if !c.Connected() {
+		return nil, ErrNotConn
+	}
+
+	if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeatureProgressiveCalls) {
+		return nil, ErrProgCallNotSupportedByRouter
+	}
+
+	if options == nil {
+		options = wamp.Dict{}
+	}
+
+	options[wamp.OptProgress] = true
+
+	// If caller is willing to receive progressive results, create a channel to
+	// receive these on.  Then, start a goroutine to receive progressive
+	// results and call the callback for each.
+	var progChan chan *wamp.Result
+	var progDone chan struct{}
+	if progcb != nil {
+		options[wamp.OptReceiveProgress] = true
+		progChan = make(chan *wamp.Result)
+		progDone = make(chan struct{})
+		go func() {
+			for result := range progChan {
+				progcb(result)
 			}
+			close(progDone)
+		}()
+	}
+
+	if finalrescb == nil {
+		return nil, errors.New("no result handler provided")
+	}
+
+	id := c.idGen.Next()
+	c.expectReply(id)
+	message := &wamp.Call{
+		Request:   id,
+		Procedure: wamp.URI(procedure),
+		Options:   options,
+	}
+
+	err := c.prepareCallPayloadMessage(message, options, args, kwargs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.sess.Send(message)
+
+	// As we can not wait for final result and must return control to the client code
+	// we wait for result in goroutine
+	go func() {
+		// Wait to receive RESULT message.
+		msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)
+
+		// Finish handling any remaining progressive results before returning the
+		// final result.
+		if progcb != nil {
+			close(progChan)
+			<-progDone
+		}
+
+		if err != nil {
```

Comment:
You could `defer func() { if err != nil { finalrescb(nil, err) } }()` earlier, just after the call to `waitForReplyWithCancel` to ensure that the callback is called, this would also simplify the logic below.

Reply:
Hmm... But if I make it through the defer — I still need to check if its err'ed to return before the switch. So I don't see any profit. Or do you mean something else?

Reply:
Just that you can have this code written once. I will re-read the diff with fresh eyes in the morning.

Reply:
Just that you can have this code written once. I will re-read the diff with fresh eyes in the morning.